sklearn/linear_model/sag.py

```diff
@@ -0,0 +1,327 @@
+"""Solvers for Ridge and LogisticRegression using SAG algorithm"""
+
+# Authors: Danny Sullivan <dbsullivan23@gmail.com>
+#          Tom Dupre la Tour <tom.dupre-la-tour@m4x.org>
+#
+# Licence: BSD 3 clause
+
+import numpy as np
+import scipy.sparse as sp
+import warnings
+
+from ..utils import ConvergenceWarning
+from ..utils.fixes import astype
+from ..utils.seq_dataset import ArrayDataset, CSRDataset
+from .sag_fast import LogLoss, SquaredLoss
+from .sag_fast import sag_sparse, get_auto_eta
+
+MAX_INT = np.iinfo(np.int32).max
+
+"""For sparse data intercept updates are scaled by this decay factor to avoid
+intercept oscillation."""
+SPARSE_INTERCEPT_DECAY = 0.01
+
+
+def make_dataset(X, y, sample_weight, random_state):
+    # check which type of Sequential Dataset is needed
+    y = astype(y, dtype=np.float64, copy=False)
+
+    if sp.issparse(X):
+        dataset = CSRDataset(X.data, X.indptr, X.indices,
+                             y, sample_weight,
+                             seed=random_state.randint(MAX_INT))
+        intercept_decay = SPARSE_INTERCEPT_DECAY
+    else:
+        dataset = ArrayDataset(X, y, sample_weight,
+                               seed=random_state.randint(MAX_INT))
+        intercept_decay = 1.0
+
+    return dataset, intercept_decay
+
+
+def sag_ridge(X, y, sample_weight=None, alpha=1e-4, max_iter=1000, tol=0.001,
+              verbose=0, random_state=None):
+    """SAG solver for Ridge regression
+
+    SAG stands for Stochastic Average Gradient: the gradient of the loss is
+    estimated each sample at a time and the model is updated along the way with
+    a constant learning rate. The inspiration for SAG comes from "Minimizing
+    Finite Sums with the Stochastic Average Gradient" by Mark Schmidt,
+    Nicolas Le Roux, and Francis Bach. 2013. <hal-00860051>
+    https://hal.inria.fr/hal-00860051/PDF/sag_journal.pdf
+
+    IMPORTANT NOTE: Ridge regressor with 'sag' solver and models from
+    linear_model in general depend on columns that are on the same scale.
```

Comment:
"depend" is not clear. Say that it helps SAG to converge faster. It's not a requirement.